### PR TITLE
Add Strawberry GraphQL integration

### PR DESCRIPTION
**Pull request by sentrivana**
Capture GraphQL errors and spans when using [Strawberry](https://strawberry.rocks).

This is a port of the existing [Sentry tracing extension in Strawberry](https://github.com/strawberry-graphql/strawberry/blob/9c946a2d4f56b81dfb0e58e4a9ed2aa89adfc1c8/strawberry/extensions/tracing/sentry.py) with added error monitoring.

Example event in Sentry: <tbd>

Strawberry is one of three integrations for https://github.com/getsentry/sentry-python/issues/2257, split from https://github.com/getsentry/sentry-python/pull/2381.

### CODE REVIEW

**Comment by sentrivana** [comment]
Hey @patrick91, here's a shot at porting the 🍓 Sentry tracing extension to the Sentry SDK and adding error monitoring. Any feedback is welcome!

Some notes:

- The integration, if enabled, will automatically remove the existing Strawberry Sentry tracing extension, if present, to avoid conflicts.
- The error monitoring will only work if folks don't overwrite the default [process_result](https://github.com/strawberry-graphql/strawberry/blob/9c946a2d4f56b81dfb0e58e4a9ed2aa89adfc1c8/strawberry/http/__init__.py#L18-L26). This is not optimal but it was the most straightforward place to hook into.
- I've made some changes to the tracing extension, mostly renaming things (e.g. span `op`s) to comply with [our](https://develop.sentry.dev/sdk/performance/span-operations/) and/or [OTel](https://opentelemetry.io/docs/specs/otel/trace/semantic_conventions/instrumentation/graphql/) conventions. Also changed all `span.set_tag`s to `span.set_data` since the SDK ideally shouldn't be setting span tags.

**Comment by patrick91** [comment]
we spoke on discord about this, I'm very excited to see this landing! From our side we'll add some code to make it easier to hook into the process result code (so users can still override it)

I'll also make sure to have an integration test on our side once this lands, so we don't break the integration in future ☺️

**Comment by patrick91** [comment]
Made the PR here: https://github.com/strawberry-graphql/strawberry/pull/3127 @sentrivana let me know if that looks ok to you!

**Comment by sentrivana** [comment]
Thanks @patrick91 for being awesome and helping us integrate with Strawberry better! This PR now utilizes a new hook released with Strawberry 0.209.5.

File: tests/integrations/strawberry/test_strawberry_py3.py [hunk 1]
```diff
@@ -0,0 +1,894 @@
...
+    events = capture_events()
+
+    schema = strawberry.Schema(Query)
+
+    async_app = FastAPI()
+    async_app.include_router(GraphQLRouter(schema), prefix="/graphql")
+
+    query = "query ErrorQuery { error }"
+    client = TestClient(async_app)
+    client.post("/graphql", json={"query": query, "operationName": "ErrorQuery"})
```
Comment (szokeasaurusrex) [inline tests/integrations/strawberry/test_strawberry_py3.py:364]: Perhaps this logic (and the equivalent logic for the sync case) could be factored out into a fixture, so that we can use the same test functions for both sync and async and have each test parameterized by the logic for constructing the async and sync apps and queries

This comment also applies to most other tests in this file

**Review by szokeasaurusrex — COMMENTED** [review]
I have suggested a few changes. But, generally looks good!

File: sentry_sdk/integrations/strawberry.py [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+                    async_execution
+                )
+            )
+        self.async_execution = async_execution
+
+    @staticmethod
+    def setup_once():
+        # type: () -> None
+        installed_packages = _get_installed_modules()
+        version = parse_version(installed_packages["strawberry-graphql"])
```
[the remark this answers is not in the document]
Comment (patrick91) [reply sentry_sdk/integrations/strawberry.py:59]: unfortunately we don't have it :(

[the remark this answers is not in the document]
Comment (sentrivana) [reply sentry_sdk/integrations/strawberry.py:59]: I'd like to move to using `_get_installed_modules` to get the version in all our integrations eventually since `importlib.metadata` (where it takes the data from) is generic, works for all pypi packages, and we're not dependent on a usually undocumented constant that, if moved, would break the whole integration.

File: sentry_sdk/integrations/strawberry.py [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+
+def _patch_views():
+    # type: () -> None
+    old_handle_errors_async = async_base_view.AsyncBaseHTTPView._handle_errors
+    old_handle_errors_sync = sync_base_view.SyncBaseHTTPView._handle_errors
+
+    def _sentry_patched_handle_errors_async(self, errors, response_data):
+        # type: (Any, List[GraphQLError], GraphQLHTTPResponse) -> None
+        old_handle_errors_async(self, errors, response_data)
+        _sentry_patched_handle_errors(self, errors, response_data)
```
[the remark this answers is not in the document]
Comment (sentrivana) [reply sentry_sdk/integrations/strawberry.py:310]: It's a [sync method](https://github.com/strawberry-graphql/strawberry/blob/a4aa93e6be75c806457f97c7a60bcb3d446ca422/strawberry/http/async_base_view.py#L149). It's just a method of the AsyncBaseHTTPView class (as opposed to SyncBaseHTTPView), hence the name here. Maybe I can come up with a less confusing name.

File: sentry_sdk/consts.py [hunk 4]
```diff
@@ -170,6 +170,13 @@ class OP:
     FUNCTION = "function"
     FUNCTION_AWS = "function.aws"
     FUNCTION_GCP = "function.gcp"
+    GRAPHQL_EXECUTE = "graphql.execute"
+    GRAPHQL_MUTATION = "graphql.mutation"
+    GRAPHQL_PARSE = "graphql.parse"
+    GRAPHQL_RESOLVE = "graphql.resolve"
+    GRAPHQL_SUBSCRIPTION = "graphql.subscription"
+    GRAPHQL_QUERY = "graphql.query"
+    GRAPHQL_VALIDATE = "graphql.validate"
```
[the remark this answers is not in the document]
Comment (sentrivana) [reply sentry_sdk/consts.py:179]: Good catch, will do!

[the remark this answers is not in the document]
Comment (sentrivana) [reply sentry_sdk/consts.py:179]: https://github.com/getsentry/develop/pull/1049

File: sentry_sdk/integrations/strawberry.py [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+            op=OP.GRAPHQL_PARSE, description="parsing"
+        )
+
+        yield
+
+        self.parsing_span.finish()
+
+    def should_skip_tracing(self, _next, info):
+        # type: (Callable[[Any, GraphQLResolveInfo, Any, Any], Any], GraphQLResolveInfo) -> bool
+        return should_skip_tracing(_next, info)
```
[the remark this answers is not in the document]
Comment (sentrivana) [reply sentry_sdk/integrations/strawberry.py:207]: Will do!

File: sentry_sdk/integrations/strawberry.py [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+            span.set_data("graphql.parent_type", info.parent_type.name)
+            span.set_data("graphql.field_path", field_path)
+            span.set_data("graphql.path", ".".join(map(str, info.path.as_list())))
+
+            result = _next(root, info, *args, **kwargs)
+
+            if isawaitable(result):
+                result = await result
+
+            return result
```
[the remark this answers is not in the document]
Comment (sentrivana) [reply sentry_sdk/integrations/strawberry.py:234]: Will do!

File: sentry_sdk/integrations/strawberry.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sentrivana) [reply sentry_sdk/integrations/strawberry.py]: Changed to `old_async_view_handle_errors`.

File: tests/integrations/strawberry/test_strawberry_py3.py [hunk 7]
```diff
@@ -0,0 +1,894 @@
...
+    sentry_init(integrations=[StrawberryIntegration(async_execution=True)])
+
+    with mock.patch(
+        "sentry_sdk.integrations.strawberry._get_installed_modules",
+        return_value={"flask": "2.3.3"},
+    ):
+        # actual installed modules should not matter, the explicit option takes
+        # precedence
+        schema = strawberry.Schema(Query)
+        assert SentryAsyncExtension in schema.extensions
```
[the remark this answers is not in the document]
Comment (sentrivana) [reply tests/integrations/strawberry/test_strawberry_py3.py:61]: There's a test case for that specifically further down 👍🏻

File: tests/integrations/strawberry/test_strawberry_py3.py [hunk 8]
```diff
@@ -0,0 +1,894 @@
...
+    assert SentryTracingExtension not in schema.extensions
+    assert SentryAsyncExtension in schema.extensions
+
+
+def test_replace_existing_sentry_sync_extension(sentry_init):
+    sentry_init(integrations=[StrawberryIntegration()])
+
+    schema = strawberry.Schema(Query, extensions=[SentryTracingExtensionSync])
+    assert SentryTracingExtensionSync not in schema.extensions
+    assert SentrySyncExtension in schema.extensions
```
[the remark this answers is not in the document]
Comment (sentrivana) [reply tests/integrations/strawberry/test_strawberry_py3.py:112]: Will do!

File: tests/integrations/strawberry/test_strawberry_py3.py (hunk 1 above)
Comment (sentrivana) [reply tests/integrations/strawberry/test_strawberry_py3.py]: Updated to use fixtures!

File: sentry_sdk/integrations/strawberry.py [hunk 9]
```diff
@@ -0,0 +1,401 @@
...
+
+                    if execution_context.variables:
+                        request_data["data"]["variables"] = execution_context.variables
+                    if execution_context.operation_name:
+                        request_data["data"][
+                            "operationName"
+                        ] = execution_context.operation_name
+
+            elif event.get("request", {}).get("data"):
+                del event["request"]["data"]
```
[the remark this answers is not in the document]
Comment (sentrivana) [reply sentry_sdk/integrations/strawberry.py]: Made one slight change to it though: https://github.com/getsentry/sentry-python/pull/2393/commits/783152002244268784cab9b730722f1f0aac1e4e So that, if for some reason `event["request"]` is there and is `None`, we don't get an `AttributeError`.

**Comment by sentrivana** [comment]
Hope I've addressed everything @szokeasaurusrex -- please have another look when you have time.

File: sentry_sdk/integrations/strawberry.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (szokeasaurusrex) [reply sentry_sdk/integrations/strawberry.py:59]: Okay, fair enough @sentrivana. However, in that case, I think maybe it would be worthwhile to extract this logic to a separate method that would retrieve and parse the version of a given package in one call. Basically, my idea would be to replace the above code with something like the following:
```python
version = _get_version_of_module("strawbery-graphql")
```

[the remark this answers is not in the document]
Comment (sentrivana) [reply sentry_sdk/integrations/strawberry.py:59]: That's a good idea. Since we need this in multiple places this is a bit out of scope for this PR, so I've opened a separate PR: https://github.com/getsentry/sentry-python/pull/2429 (I will add strawberry there as soon as this is merged).

File: sentry_sdk/integrations/strawberry.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (szokeasaurusrex) [reply sentry_sdk/integrations/strawberry.py]: I have thought about this a bit more, and I think I would really prefer to switch to the try/except syntax, since I think it is much more readable for people looking at this code for the first time.

With the try/except block, the code reads in plain English as "If we want to send PII, send it, otherwise ensure we don't send any request data." It is easy to see what the code is doing.

With the `elif` statement, the code reads in plain English as "If we want to send PII, send it, otherwise check whether the event has a request (treat the request as an empty dictionary if it doesn't) and check whether the event's request has any data, and if it does, delete the event request data." In this situation, the code's objective is less obvious; only after thoroughly reading the contents of the elif condition and the statement it contains is it clear that the code here is designed to ensure the request has no data when we send the event.

Although the `elif` version is more concise, this concision comes at the expense of readability and understandability, so I think the try/except version would be preferable here.

File: tests/integrations/strawberry/test_strawberry_py3.py [hunk 10]
```diff
@@ -0,0 +1,650 @@
...
+@pytest.mark.parametrize(
+    "client_factory,async_execution,framework_integrations",
+    (
+        (
+            "async_app_client_factory",
+            True,
+            [FastApiIntegration(), StarletteIntegration()],
+        ),
+        ("sync_app_client_factory", False, [FlaskIntegration()]),
+    ),
```
Comment (szokeasaurusrex) [inline tests/integrations/strawberry/test_strawberry_py3.py:146]: It seems like you are using this same combination of parameters for most of the tests. So I would suggest that you somehow store these default parameters in a single variable, then reuse this everywhere. You could even I think define a custom decorator for all the tests like so:
```python
parameterize_strawberry_test = pytest.mark.parametrize(
    "client_factory,async_execution,framework_integrations",
    (
        (
            "async_app_client_factory",
            True,
            [FastApiIntegration(), StarletteIntegration()],
        ),
        ("sync_app_client_factory", False, [FlaskIntegration()]),
    ),
)
```
And then, you can decorate each of your tests with `@parameterize_strawberry_test`

File: sentry_sdk/integrations/strawberry.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (sentrivana) [reply sentry_sdk/integrations/strawberry.py]: Ok, changed: https://github.com/getsentry/sentry-python/pull/2393/commits/2ccd1eb3125decec6e58689b7c203632e44a59ff
(The `TypeError` is there for the `event["request"] == None` case -- it'd throw that instead of the `AttributeError` I mentioned above.)

**Review by szokeasaurusrex — COMMENTED** [review]
Added one comment to your tests, also please address my reply to our `try`/`except` thread. Then, the code looks good to me! @antonpirker should review before merge to double-check that I didn't miss anything

File: tests/integrations/strawberry/test_strawberry_py3.py (hunk 10 above)
Comment (sentrivana) [reply tests/integrations/strawberry/test_strawberry_py3.py]: [Done!](https://github.com/getsentry/sentry-python/pull/2393/commits/ed52750d0e764e0755137ed243586a2faf420b11)

**Review by antonpirker — APPROVED** [review]
🔥 🚀
